cqrs-codegen/impl/src/event_sourced.rs

```diff
@@ -0,0 +1,144 @@
+
+/// Name of the attribute, used by [`cqrs::EventSourced`].
+const ATTR_NAME: &str = "event_sourced";
+
+/// Implements [`crate::event_sourced_derive`] macro expansion.
+pub fn derive(input: syn::DeriveInput) -> Result<TokenStream> {
+    util::derive(input, TRAIT_NAME, derive_struct, derive_enum)
+}
+
+/// Reports error if [`crate::event_sourced_derive`] macro applied to structs.
+fn derive_struct(input: syn::DeriveInput) -> Result<TokenStream> {
+    match input.data {
+        syn::Data::Struct(data) => Err(Error::new(
+            data.struct_token.span(),
+            format!("Structs are not supported for deriving {}", TRAIT_NAME),
+        )),
+        _ => unreachable!(),
+    }
+}
+
+/// Implements [`crate::event_sourced_derive`] macro expansion for enums
+/// via [`synstructure`].
+fn derive_enum(input: syn::DeriveInput) -> Result<TokenStream> {
+    let meta = util::get_nested_meta(&input.attrs, ATTR_NAME)?;
+
+    let aggregate = parse_event_sourced_aggregate(&meta)?;
+    let aggregate: syn::Path = syn::parse_str(&aggregate)?;
+
+    let mut structure = Structure::try_new(&input)?;
+    util::assert_all_enum_variants_have_single_field(&structure, TRAIT_NAME)?;
+
+    structure.binding_name(|field, _| {
+        field.ident.as_ref().map_or_else(
+            || syn::Ident::new("ev", proc_macro2::Span::call_site()),
+            |ident| ident.clone(),
+        )
+    });
+
+    let body = structure.each(|binding_info| {
+        let ev = &binding_info.binding;
+        quote!(self.apply(#ev))
+    });
+
+    let type_name = &input.ident;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+    let type_params: HashSet<_> = input
+        .generics
+        .params
+        .iter()
+        .filter_map(|generic_param| match generic_param {
+            syn::GenericParam::Type(type_param) => Some(&type_param.ident),
+            _ => None,
+        })
+        .collect();
+
+    let iter = structure
+        .variants()
+        .iter()
+        .map(|variant| variant.ast().fields.iter())
+        .flatten();
+
+    let mut where_clause = where_clause.cloned();
+
+    for field in iter {
+        if let Some(ty) = util::get_type_if_type_parameter_used_in_type(&type_params, &field.ty) {
+            let predicates = &mut where_clause
+                .get_or_insert_with(|| syn::WhereClause {
+                    where_token: <syn::Token![where]>::default(),
+                    predicates: Punctuated::new(),
+                })
+                .predicates;
+
+            predicates.push(syn::parse2(quote!(#aggregate: ::cqrs::EventSourced<#ty>))?);
+            predicates.push(syn::parse2(quote!(#ty: ::cqrs::Event))?);
+        }
+    }
+
+    Ok(quote! {
+        #[automatically_derived]
+        impl#impl_generics ::cqrs::EventSourced<#type_name#ty_generics> for #aggregate #where_clause {
+            #[inline(always)]
+            fn apply(&mut self, ev: &#type_name#ty_generics) {
```

Comment:
@ffuugoo matching over `enum` with a lot of variants is not that trivial to be strictly inlined all the time. Here is the place where compiler should decide this, like we do in `derive(Event)` for `enum`s.